src/ApacheOrcDotNet.OptimizedReaderTest.App/TradeDataSource.cs

```diff
@@ -0,0 +1,127 @@
+            => new TimeRangeReader(_reader, _souce, _symbol, startTime, endTime);
+    }
+
+    public class TimeRangeReader
+    {
+        private readonly Reader _reader;
+        private readonly string _souce;
+        private readonly string _symbol;
+        private readonly TimeSpan _startTime;
+        private readonly TimeSpan _endTime;
+        private readonly BaseColumnBuffer<string> _sourceColumnBuffer;
+        private readonly BaseColumnBuffer<string> _symbolColumnBuffer;
+        private readonly BaseColumnBuffer<decimal?> _timeColumnBuffer;
+        private readonly BaseColumnBuffer<decimal?> _priceColumnBuffer;
+        private readonly BaseColumnBuffer<long?> _sizeColumnBuffer;
+        private readonly Dictionary<int, List<int>> _filters = new();
+
+        public TimeRangeReader(Reader reader, string source, string symbol, TimeSpan startTime, TimeSpan endTime)
+        {
+            _reader = reader;
+            _souce = source;
+            _symbol = symbol;
+            _startTime = startTime;
+            _endTime = endTime;
+
+            var sourceColumn = _reader.GetColumn("source");
+            var symbolColumn = _reader.GetColumn("symbol");
+            var timeColumn = _reader.GetColumn("time");
+            var priceColumn = _reader.GetColumn("price");
+            var sizeColumn = _reader.GetColumn("size");
+
+            _sourceColumnBuffer = reader.CreateStringColumnBuffer(sourceColumn);
+            _symbolColumnBuffer = reader.CreateStringColumnBuffer(symbolColumn);
+            _timeColumnBuffer = reader.CreateDecimalColumnBuffer(timeColumn);
+            _priceColumnBuffer = reader.CreateDecimalColumnBuffer(priceColumn);
+            _sizeColumnBuffer = reader.CreateIntegerColumnBuffer(sizeColumn);
+
+            var sourceFilterValues = FilterValues.CreateFromString(min: _souce, max: _souce);
+            var symbolFilterValues = FilterValues.CreateFromString(min: _symbol, max: _symbol);
+            var timeFilterValues = FilterValues.CreateFromTime(min: _startTime, max: _endTime);
+
+            var stripeIds = _reader.FilterStripes(sourceColumn, sourceFilterValues);
+            stripeIds = _reader.FilterStripes(stripeIds, symbolColumn, symbolFilterValues);
+            stripeIds = _reader.FilterStripes(stripeIds, timeColumn, timeFilterValues);
+
+            foreach (var stripeId in stripeIds)
+            {
+                var rowGroupIndexes = _reader.FilterRowGroups(stripeId, sourceColumn, sourceFilterValues);
+                rowGroupIndexes = _reader.FilterRowGroups(rowGroupIndexes, stripeId, symbolColumn, symbolFilterValues);
+                rowGroupIndexes = _reader.FilterRowGroups(rowGroupIndexes, stripeId, timeColumn, timeFilterValues);
+
+                _filters.Add(stripeId, rowGroupIndexes.ToList());
+            }
+        }
+
+        public int ApproxRowCount
+        {
+            get
+            {
+                var rowGroups = _filters.Values.Sum(v => v.Count);
```

Comment:
I like this.  I wasn't sure it would work to approximate the number of rows but this looks like it'll work perfectly.